gzhttp/transport.go

```diff
@@ -114,3 +131,32 @@ func lower(b byte) byte {
 	}
 	return b
 }
+
+// gzipReader wraps a response body so it can lazily
+// call gzip.NewReader on the first call to Read
+type zstdReader struct {
+	body io.ReadCloser // underlying HTTP/1 response body framing
+	zr   *zstd.Decoder // lazily-initialized gzip reader
+	zerr error         // any error from zstd.NewReader; sticky
+}
+
+func (zr *zstdReader) Read(p []byte) (n int, err error) {
+	if zr.zr == nil {
+		if zr.zerr == nil {
+			zr.zr, zr.zerr = zstd.NewReader(zr.body, zstd.WithDecoderLowmem(true), zstd.WithDecoderMaxWindow(32<<20))
+		}
+		if zr.zerr != nil {
+			return 0, zr.zerr
+		}
+	}
+
+	return zr.zr.Read(p)
+}
+
+func (zr *zstdReader) Close() error {
+	if zr.zr != nil {
+		zr.zr.Close()
+		zr.zr = nil
+	}
+	return zr.body.Close()
```

Comment:
You may want to add a finalizer to ensure this close is called. net/http has bugs where sometimes close is called twice, sometimes it is not closed. See:

https://go-review.googlesource.com/c/go/+/124315
https://go-review.googlesource.com/c/go/+/125636
https://go-review.googlesource.com/c/go/+/125635

I lost interest in pursuing these once the PRs languished for over a year.

Reply:
@twmb After #498 there now is a stateless decoder, which means we can safely (re)use it, even if close isn't called.